javascript/src/index.js

```diff
@@ -72,160 +73,356 @@ function preventUnintendedZoomOnScroll(map) {
-      .on("layeradd layerremove", function(e) {
-        // If the layer that's coming or going is a group we created, tell
-        // the server.
-        if (map.layerManager.getGroupNameFromLayerGroup(e.layer)) {
-          // But to avoid chattiness, coalesce events
-          if (groupTimerId) {
-            clearTimeout(groupTimerId);
-            groupTimerId = null;
+  factory: function(el, width, height) {
+
+    let map = null;
+
+    return {
+
+      // we need to store our map in our returned object.
+      getMap: function() {
+        return map;
+      } ,
+
+      renderValue: function(data) {
+
+        // Create an appropriate CRS Object if specified
+
+        if(data && data.options && data.options.crs) {
+
+          let crs = L.CRS.EPSG3857; // Default Spherical Mercator
+          let crsOptions = data.options.crs;
+
+          switch(crsOptions.crsClass) {
+          case "L.CRS.EPSG3857":
+            crs = L.CRS.EPSG3857;
+            break;
+          case "L.CRS.EPSG4326":
+            crs = L.CRS.EPSG4326;
+            break;
+          case "L.CRS.EPSG3395":
+            crs = L.CRS.EPSG3395;
+            break;
+          case "L.CRS.Simple":
+            crs =L.CRS.Simple;
+            break;
+          case "L.Proj.CRS":
+            crs = new Proj4Leaflet.CRS(crsOptions.code, crsOptions.proj4def,
+                             crsOptions.options);
+            break;
+          case "L.Proj.CRS.TMS":
+            crs = new Proj4Leaflet.CRS.TMS(crsOptions.code, crsOptions.proj4def,
+                                 crsOptions.projectedBounds, crsOptions.options);
+            break;
           }
-          groupTimerId = setTimeout(function() {
-            groupTimerId = null;
-            Shiny.onInputChange(map.id + "_groups",
-              map.layerManager.getVisibleGroups());
-          }, 100);
+          data.options.crs = crs;
         }
-      });
-
-    return map;
-  },
-  renderValue: function(el, data, map) {
-    return this.doRenderValue(el, data, map);
-  },
-  doRenderValue: function(el, data, map) {
-    // Leaflet does not behave well when you set up a bunch of layers when
-    // the map is not visible (width/height == 0). Popups get misaligned
-    // relative to their owning markers, and the fitBounds calculations
-    // are off. Therefore we wait until the map is actually showing to
-    // render the value (we rely on the resize() callback being invoked
-    // at the appropriate time).
-    //
-    // There may be an issue with leafletProxy() calls being made while
-    // the map is not being viewed--not sure what the right solution is
-    // there.
-    if (el.offsetWidth === 0 || el.offsetHeight === 0) {
-      map.leafletr.pendingRenderData = data;
-      return;
-    }
-    map.leafletr.pendingRenderData = null;
 
-    // Merge data options into defaults
-    let options = $.extend({ zoomToLimits: "always" }, data.options);
+        map = L.map(el, data.options);
```

Comment:
will do. I need to read up on htmlwidgets code to truly understand what happens when, and scenarios where there will already be an instance of map.